contracts/core/MarginRequirements.sol

```diff
@@ -0,0 +1,185 @@
+
+        initialMargin[keccak256(abi.encode(_underlying, _collateralAsset, _isPut))][_account] = _initialMargin;
+    }
+
+    /**
+     * @notice sets the maintenance margin absolute amount
+     * @dev can only be called by keeper
+     * @param _vaultID id of the vault
+     * @param _maintenanceMargin maintenance margin absolute amount
+     */
+    function setMaintenanceMargin(uint256 _vaultID, uint256 _maintenanceMargin) external onlyKeeper {
+        require(_maintenanceMargin > 0, "MarginRequirements: initial margin cannot be 0");
+
+        maintenanceMargin[_vaultID] = _maintenanceMargin;
+    }
+
+    /************************************************
+     *  MARGIN OPERATIONS
+     ***********************************************/
+
+    /**
+     * @notice checks if there is enough collateral to mint the desired amount of otokens
+     * @param _account account address
+     * @param _notional order notional amount
+     * @param _underlying underlying asset address
+     * @param _isPut option type the vault is selling
+     * @param _collateralAsset collateral asset address
+     * @param _collateralAmount collateral amount
+     * @return boolean value stating whether there is enough collateral to mint
+     */
+    function checkMintCollateral(
+        address _account,
+        uint256 _notional,
+        address _underlying,
+        bool _isPut,
+        uint256 _collateralAmount,
+        address _collateralAsset
+    ) external view returns (bool) {
+        uint256 collateralDecimals = uint256(ERC20Interface(_collateralAsset).decimals());
+
+        uint256 initialMarginRequired = initialMargin[keccak256(abi.encode(_underlying, _collateralAsset, _isPut))][
+            _account
+        ];
+
+        return
+            _notional.mul(initialMarginRequired).mul(10**collateralDecimals).mul(1e8) <
+            _collateralAmount.mul(oracle.getPrice(_collateralAsset)).mul(100e2).mul(1e6);
+    }
+
+    /**
+     * @notice checks if there is enough collateral to withdraw the desired amount
+     * @param _account account address
+     * @param _notional order notional amount
+     * @param _withdrawAmount desired amount to withdraw
+     * @param _otokenAddress otoken address
+     * @param _vaultID id of the vault
+     * @param _vault vault struct
+     * @return boolean value stating whether there is enough collateral to withdraw
+     */
+    function checkWithdrawCollateral(
+        address _account,
+        uint256 _notional,
+        uint256 _withdrawAmount,
+        address _otokenAddress,
+        uint256 _vaultID,
+        MarginVault.Vault memory _vault
+    ) external view returns (bool) {
+        uint256 collateralDecimals = uint256(ERC20Interface(_vault.collateralAssets[0]).decimals());
+
+        return
+            _notional.mul(_getInitialMargin(_otokenAddress, _account)).mul(10**collateralDecimals).mul(1e8) <
+            (_vault.collateralAmounts[0].sub(_withdrawAmount).sub(maintenanceMargin[_vaultID]))
```

Comment:
its possible withdraw amount + maintenance margin > _vault.collateralAmounts[0] so this sub would overflow and revert in the assert in withdrawCollateral in OTC. maybe can do check in withdrawCollateral further upstream that the collateral amount is sufficient to cover the withdraw amount and maintenance margin in the first place

Reply:
https://github.com/ribbon-finance/GammaProtocol-OTC/pull/2/commits/2f6580d3737f631b016830a29df8f31113d3dc85

Reply:
should just return false here, not need to revert (we revert in OTC)

Reply:
makes sense https://github.com/ribbon-finance/GammaProtocol-OTC/pull/2/commits/9fd3dc2caf6b308bc4f9c32e00d034102924aed6